Anomaly gear: Polymorphic Field Inverter

Jacquerel opened the pull request:
## About The Pull Request

https://www.youtube.com/watch?v=5bGrqJkpbLw

Adds a new device you can research called the "Polymorphic Field Inverter".
It requires Advanced Biotechnology and Anomalous Research to unlock.

Completing that research lets you print out a fancy belt. Using it for 5 seconds on a mob (dead or alive) which is organic in nature and isn't some kind of human or megafauna (monkeys are fine) will store a copy of its DNA. Scanning another creature will replace the original stored DNA. Storing DNA does absolutely nothing unless you have also inserted an active Bioscrambler core.

If the item is activated with a Bioscrambler core then equipping it grants you the ability to transform into the same kind of creature as the one you scanned (and back). As in spells such as Wildshape you share health with your transformed form and if you die while transformed then you die, so be careful when turning into something like a cockroach which has one hit point.

It has an associated experiment which makes the research free, which is to scan two humans with (non-prosthetic) organs that they didn't have when the round started. You can accomplish this through surgery, DNA infusion, or a bioscrambler anomaly. You can do it on some monkeys if nobody is cooperating with you.
I copied a teeny bit of this code from #73969 although I didn't actually end up needing most of it.

These are absolutely codersprites so if anyone wanted to make cooler ones, let me know.

## Why It's Good For The Game

Adds some fun tech linked to an anomaly core which has literally no in-game purpose.
There's probably some neat sandbox interactions you can get out of this.

## Changelog

:cl:
add: Adds a new technology which utilises Bioscrambler anomaly cores to allow you to transform into animals.
/:cl:

Jacquerel committed:
Heartlessly destroys soul dividers

Jacquerel committed:
it works

Jacquerel committed:
Add tech node

Jacquerel committed:
Visuals

Jacquerel committed:
Add discount experiment

Jacquerel committed:
Funny text

Pickle-Coding commented:
If someone used this on the paper wizard, would they keep the AI that basically instantly kills everyone near them?

Jacquerel committed:
Fine have an unused default

CCC23 commented:
> > If someone used this on the paper wizard, would they keep the AI that basically instantly kills everyone near them?
> 
> It'll have the abilities of a paper wizard but not be controlled by an AI. To acheive this you will need to create this device, go to space, then stand next to the paper wizard for five seconds while it attempts to kill you, as it deletes itself upon death. Then put a bioscrambler core in it. If you want to do that as a nonantagonist just so you can get banned for flagrant collateral damage that seems fine, if you want to do it as an antagonist then you're taking a pretty big risk for a small payoff.

you underestimate the stations retardation also does this include elites from mining?

newfren commented:
I bet you could facetank either the hierophant or the blood drunk miner for 5 seconds which would be very funny on station

otherwise this rules, more uses for cores that aren't just the armour is great. would this trigger the requirement for the lepton violet from #77277 ?

atlasle commented:
this is just fun as hell, don't forget to update your changelog

Jacquerel commented:
I think that this does not go through the same code path which unlocks the Lepton Violet in that PR, as it's not a permanent change I don't know that it should either.

I believe that Lavaland Elites have the same flags restricting copying them as the bosses do.

distributivgesetz commented on code/modules/research/designs/misc_designs.dm:
```diff
@@ -501,9 +496,7 @@
-/////////////////////////////////////////
-////////////Janitor Designs//////////////
-/////////////////////////////////////////
+// anitor Designs
```
the j

Jacquerel replied:
```suggestion
// Janitor Designs
```

Jacquerel committed:
Update code/modules/research/designs/misc_designs.dm

VladinXXV commented:
wheres the feature tag

nukularpower commented:
sweet, love to see more uses for anomalies

Time-Green commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+
+	cast_on.balloon_alert(cast_on, "transforming...")
+	if (!do_after(cast_on, delay = channel_time, target = cast_on))
+		animate(cast_on, transform = matrix(), time = 0.25, easing = SINE_EASING)
```
```suggestion
		animate(cast_on, transform = matrix(), time = 0.25 SECONDS, easing = SINE_EASING)
```

Did you intend seconds here?

Jacquerel replied:
Weirdly enough this was a mistake but actually gives the effect I wanted, so I set it to 0 instead.

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+		return
+	qdel(weapon)
+	active = TRUE
+	update_appearance(UPDATE_ICON_STATE)
+	user.update_worn_belt()
```
Use `/datum/element/update_icon_updates_onmob`

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+	if (!(target_mob.mob_biotypes & MOB_ORGANIC))
+		balloon_alert(user, "organic life only!")
+		return
+	if (isanimal_or_basicmob(target_mob))
+		if (!target_mob.compare_sentience_type(SENTIENCE_ORGANIC))
+			balloon_alert(user, "target too intelligent!")
+			return
```
Does this prevent Megafauna from being scanned

Jacquerel replied:
yes

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+/obj/item/polymorph_belt/attackby(obj/item/weapon, mob/user, params)
+	if (!istype(weapon, /obj/item/assembly/signaler/anomaly/bioscrambler))
+		return ..()
+	balloon_alert(user, "inserting...")
+	if (!do_after(user, delay = 3 SECONDS, target = src))
```
You probably want to `return TRUE` after this to prevent afterattack (regardless of if the do_after succeeds)

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+	return slot & ITEM_SLOT_BELT
+
+/obj/item/polymorph_belt/update_icon_state()
+	icon_state = active ? "polybelt" : "polybelt_inactive"
+	worn_icon_state = active ? "polybelt" : "polybelt_inactive"
```
Use `base_icon_state`

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,152 @@
+	animate(owner, transform = matrix() * 0.1, time = 0, easing = JUMP_EASING)
+	animate(transform = matrix(), time = 0.25 SECONDS, easing = SINE_EASING)
+
+/datum/action/cooldown/spell/shapeshift/polymorph_belt/Remove(mob/living/remove_from)
+	unshift_owner()
+	return ..()
```
I'm surprised this works considering it `Remove`s then `Grant`s on cast (to the shapechange mob)

(Though this should probably be default behavior of shapechange)

Jacquerel replied:
Oh you're right, I guess it works because while it is doing that, the buff has not finished being applied yet.

Shadowflame909 commented:
We are so back, staff of change dreams never die

Jacquerel committed:
Make shapeshift actions unshift you by default when removed
Use update_icon_updates_onmob element
Use base_icon_state
Interrupt attack chain
Fix my animation

Boopideedoo commented:
So if I was to use this on as xenomorph that was a Charlie stations anymore would I just become a useless mob like the Charlie station mobs? Or would I gain powers to lay resin and whatnot?

Jacquerel commented:
> So if I was to use this on a xenomorph that was a Charlie station xenomorph would I just become a useless mob like the Charlie station mobs? Or would I gain powers to lay resin and whatnot?

You will become whatever you copied, so if the charlie xeno mobs don't have resin you won't either.

MrMelbert commented on code/modules/clothing/belts/polymorph_belt.dm:
```diff
@@ -0,0 +1,159 @@
+	/// Amount of time it takes us to transform back or forth
+	var/channel_time = 3 SECONDS
+
+/datum/action/cooldown/spell/shapeshift/Remove(mob/remove_from)
```
```suggestion
/datum/action/cooldown/spell/shapeshift/polymorph_belt/Remove(mob/remove_from)
```

Jacquerel replied:
oops 😅

Jacquerel committed:
Update code/modules/clothing/belts/polymorph_belt.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Jacquerel committed:
Don't update memories or turn into poly's ghost
